### PR TITLE
MS.CSharp Refactoring of producing lambda for dynamic operation and rewriting it

### PR DESCRIPTION
**Pull request by JonHanna**
* Tighten type of parents in symbol factory methods to those valid.

Makes the correctness of what follows clearer.

* Remove parent parameter to `SymFactory.CreateScope()`

Always null; we only ever work with one `Scope`. Let's make that explicit

* Remove `RewriteLambdaParameters`

Pulls locals that are sibling to scope into lambda expression (for hoisting) but there is never a sibling to any scope in the dynamic binder.

* Remove `ExpressionTreeRewriter.DestroyWraps`

Unwraps hoisted locals that are siblings to the scope, but we never have any.

* Remove `ExpressionTreeRewriter.currentAnonMeth`

We only ever have one bound lambda at a time; never another nested within that, so this check is unnecessary.

* Remove check `createParameters != null`

Always true. Indeed `ExpressionTreeCallRewriter.Rewrite` depends on this as it expects an `ExprBinOp`, and if this path wasn't taken we'd return an `ExprCall` instead.

* Take `VisitBOUNDLAMBDA` out of the main visiting loop.

Can only be reached from the top of an `ExpressionTreeRewriter` visit, so only call it from there, tightening parameter types.

* Tighten `VisitBoundLambda` to return `ExprBinOp`

Pass this tightened type definition through to `ExpressionTreeCallRewriter.Rewrite`, making one asserted fact there statically known.

* Remove `ExpressionKind.SequenceReverse`

No such expression is ever created.

* Create predefined type for `Func<>`

Since we're using it with every dynamic operation, it should be in the faster lookup.

* Tighten `ExprBoundLambda` ctor parameter types and assertions.

* Place lambda's return expression directly inside `ExprBoundLambda`

Since the expression is always a single block with a single return that returns a single expression, just store that single expression.

Essentially this changes the restrictions from what one can do in C# in defining an anonymous delegate to what one can do in C# in defining an expresson with `=>` syntax.

This matches what we are actually using the object for better, and removes two levels of indirection through other objects, with the possibility of errors in them.

* Remove `ExprBlock`, `ExprReturn` & `ExprStatement`

No longer used.

Remove commented-out `ExpressionKind`s along with those removed here.

* Don't add "`this`" local to scope.

Never given meaning, only ever deliberately avoided in rewrites.

When a dynamic operation uses `this`, it does so by having `this` passed-in explicitly.

* Remove `LocalVariableSymbol.fUsedInAnonMeth`

Never not set to true, so no real significance.

* Pass deferred binding to out parameter directly.

We'll always get the correct final result there, so don't replicate logic we just called into.

* Remove path in `ExpressionTreeRewriter.VisitLOCAL` for null `wrap`.

Since the only locals have just had this set to non-null, this is unreachable.

Since the code removed here is the only use of `ExprHoistedLocalExpr`, remove it too.

### CODE REVIEW

**Comment by JonHanna** [comment]
This will conflict with #25484, and that fixes an issue while this merely refactors, so closing to re-open after #25484 is dealt with.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/LocalVariableSymbol.cs [hunk 1]
```diff
@@ -14,9 +14,7 @@ internal sealed class LocalVariableSymbol : VariableSymbol
 
         public ExprWrap wrap;
 
-        public bool isThis;           // Is this the one and only <this> pointer?
         // movedToField should have iIteratorLocal set appropriately
-        public bool fUsedInAnonMeth;   // Set if the local is ever used in an anon method
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/LocalVariableSymbol.cs]: Is that comment relevant now?

**Comment by JonHanna** [comment]
Test Alpine.3.6 x64 Debug Build
(Failure at `chmod +x` https://ci3.dot.net/job/dotnet_corefx/job/master/job/alpine-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/6/consoleFull)

**Comment by JonHanna** [comment]
Test Alpine.3.6 x64 Debug Build
(Same)

**Comment by OmarTawfik** [comment]
@dotnet/dnceng FYI. this build seems to be failing forever. Is someone looking into this/an open issue about it?

**Comment by JonHanna** [comment]
Test Alpine.3.6 x64 Debug Build

**Comment by markwilkie** [comment]
From the log, it looks like the build is failing with 10:24:09   chmod: cannot access '/mnt/j/workspace/dotnet_corefx/master/alpine-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/bin/testhost/netcoreapp-Linux-Debug-x64/dotnet': No such file or directory 

Maybe some problem with testhost?  (doesn't seem too infrastructury at least as far as I can tell right now)

**Comment by JonHanna** [comment]
It was passing before the last commit, and that only deleted a comment, but I'd still like to see them all green.

**Comment by markwilkie** [comment]
@mmitche - any top of mind thoughts as to why this would be failing for Alpine only?

**Comment by mmitche** [comment]
@markwilkie Not sure, but I don't think this is infrastructure.  My guess is that the directory simply doesn't exist.

**Comment by OmarTawfik** [comment]
@markwilkie @mmitche I don't think that job ever succeeded:
https://ci3.dot.net/job/dotnet_corefx/job/master/job/alpine-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/

**Comment by mmitche** [comment]
@OmarTawfik It was added ysterday.  @janvorli Can you take a look?

**Comment by janvorli** [comment]
@mmitche I know it doesn't work, I have a PR to fix that (needs updating dotnet host version in dependency.props), but it breaks Invariant.Tests tests on all platforms. Due to that, I am going to update the CI change to not to try to build tests, which would work. We mostly need this CI job to prevent breaking Alpine native build and that would work for this purpose for now until the Invariant test issue is understood and fixed.

**Comment by JonHanna** [comment]
Test Alpine.3.6 x64 Debug Build

**Comment by JonHanna** [comment]
Test this please

**Pull request merged**